 # Notebook to get abstracts from Semantic Scholar using DOIs, PMIDs, or PMCIDs 
 
Note: This uses the filtered_data_with_identifiers.csv file which has the identifiers for the papers. This file was created using the notebook `get_identifiers_from_links.ipynb`. Please generate that file (or get it from google drive) and put it in a folder called `data` before running this notebook. This also requires an api key for semantic scholar. 

In [1]:
import pandas as pd

In [2]:
inp_df = pd.read_csv('data/filtered_data_with_identifiers.csv')
print(inp_df.shape)
inp_df.head()

(8299, 23)


Unnamed: 0  year  month                                              title  \
0           5  2018      3  Firearm Injuries Drop 20 Percent When Gun Owne...   
1          19  2018      3  Supplementation with probiotics during late pr...   
2          25  2018      3  Study finds that bee venom could be a useful p...   
3          31  2018      3  The interplay of gene flow, population size va...   
4          33  2018      3  Undisclosed Conflicts of Interests among Biome...   

  link_flair_text                   domain  score  num_comments  \
0         Biology                 nejm.org     84            22   
1          Health        journals.plos.org      8             1   
2        Medicine         ncbi.nlm.nih.gov     27             9   
3         Biology  onlinelibrary.wiley.com      6             0   
4  Social Science         ncbi.nlm.nih.gov    263            21   

   sensationalism_score  jargon_proportion  ... is_top_domain_scientific  \
0              0.530595           0.000000  ...                    False   
1              0.482136           0.314286  ...                     True   
2              0.507827           0.333333  ...                    False   
3              0.433548           0.352941  ...                     True   
4              0.511708           0.086957  ...                    False   

  is_top_domain_news  is_top_domain_repo  is_top_domain_scam  \
0              False                True               False   
1              False               False               False   
2              False                True               False   
3              False               False               False   
4              False                True               False   

   is_top_domain_unknown  is_top_domain_indecisive  is_top_domain_less_than_2  \
0                  False                     False                      False   
1                  False                     False                      False   
2                  False                     False                      False   
3                  False                     False                      False   
4                  False                     False                      False   

   label_voting_lm  label_voting_manual                        identifier  
0             repo                 repo          DOI:10.1056/NEJMc1712773  
1       scientific           scientific  DOI:10.1371/journal.pmed.1002507  
2             repo                 repo                  PMCID:PMC5793096  
3       scientific           scientific    DOI:10.1111/evo.13435/abstract  
4             repo                 repo                     PMID:29400625  

[5 rows x 23 columns]

In [3]:
# Using semantic scholar bulk API to get abstracts
import requests
import time
import os
from tqdm import tqdm


In [4]:
# TODO: Get the API key from the environment variable
SEM_SCHOLAR_API_KEY = os.getenv('SEM_SCHOLAR_API_KEY')

In [5]:
# API endpoint: https://api.semanticscholar.org/graph/v1/paper/batch
headers = {
    'x-api-key': SEM_SCHOLAR_API_KEY
}

def fetch_paper_details(ids):
    url = 'https://api.semanticscholar.org/graph/v1/paper/batch'
    params = {
        'fields': 'title,abstract'
    }
    response = requests.post(url, headers=headers, json={'ids': ids}, params=params)
    if response.status_code == 200:
        return response.json()
    else:
        print(f"Error: {response.status_code} for IDs: {ids}")
        return []


batch_size = 100
results = []
for i in tqdm(range(0, len(inp_df), batch_size)):
    batch_ids = inp_df['identifier'].iloc[i:i + batch_size].tolist()
    data = fetch_paper_details(batch_ids)
    results.extend(data)
    # If we get a result, add columns title and abstract to df
    if data:
        for j, r in enumerate(data):
            if r:
                sem_title = r.get('title', None)
                sem_abstract = r.get('abstract', None)
                if sem_title:
                    inp_df.loc[i+j, 'sem_scholar_title'] = sem_title
                if sem_abstract:
                    inp_df.loc[i+j, 'sem_scholar_abstract'] = sem_abstract         
    time.sleep(1.5)  # Max 1 request per second, adding some buffer as I got some 429 errors


100%|██████████| 83/83 [02:51<00:00,  2.06s/it]


In [6]:
inp_df.head()

Unnamed: 0  year  month                                              title  \
0           5  2018      3  Firearm Injuries Drop 20 Percent When Gun Owne...   
1          19  2018      3  Supplementation with probiotics during late pr...   
2          25  2018      3  Study finds that bee venom could be a useful p...   
3          31  2018      3  The interplay of gene flow, population size va...   
4          33  2018      3  Undisclosed Conflicts of Interests among Biome...   

  link_flair_text                   domain  score  num_comments  \
0         Biology                 nejm.org     84            22   
1          Health        journals.plos.org      8             1   
2        Medicine         ncbi.nlm.nih.gov     27             9   
3         Biology  onlinelibrary.wiley.com      6             0   
4  Social Science         ncbi.nlm.nih.gov    263            21   

   sensationalism_score  jargon_proportion  ... is_top_domain_repo  \
0              0.530595           0.000000  ...               True   
1              0.482136           0.314286  ...              False   
2              0.507827           0.333333  ...               True   
3              0.433548           0.352941  ...              False   
4              0.511708           0.086957  ...               True   

  is_top_domain_scam  is_top_domain_unknown  is_top_domain_indecisive  \
0              False                  False                     False   
1              False                  False                     False   
2              False                  False                     False   
3              False                  False                     False   
4              False                  False                     False   

   is_top_domain_less_than_2  label_voting_lm  label_voting_manual  \
0                      False             repo                 repo   
1                      False       scientific           scientific   
2                      False             repo                 repo   
3                      False       scientific           scientific   
4                      False             repo                 repo   

                         identifier  \
0          DOI:10.1056/NEJMc1712773   
1  DOI:10.1371/journal.pmed.1002507   
2                  PMCID:PMC5793096   
3    DOI:10.1111/evo.13435/abstract   
4                     PMID:29400625   

                                   sem_scholar_title  \
0  Reduction in Firearm Injuries during NRA Annua...   
1  Diet during pregnancy and infancy and risk of ...   
2                                                nan   
3                                                nan   
4  Undisclosed conflicts of interest among biomed...   

                                sem_scholar_abstract  
0  Decline in Firearm Injuries during NRA Convent...  
1  Background There is uncertainty about the infl...  
2                                                nan  
3                                                nan  
4  ABSTRACT Background: Textbooks are a formative...  

[5 rows x 25 columns]

In [7]:
# if sem scholar title or abstract is nan or text is "nan", then remove that row
inp_df = inp_df[(~inp_df['sem_scholar_title'].isnull()) & (inp_df['sem_scholar_title'] != 'nan')]
inp_df = inp_df[(~inp_df['sem_scholar_abstract'].isnull()) & (inp_df['sem_scholar_abstract'] != 'nan')]

inp_df

Unnamed: 0  year  month  \
0              5  2018      3   
1             19  2018      3   
4             33  2018      3   
6             53  2018      3   
7             57  2018      3   
...          ...   ...    ...   
8294      197738  2019     10   
8295      197742  2019     10   
8296      197760  2019     10   
8297      197779  2019     10   
8298      197784  2019     10   

                                                  title link_flair_text  \
0     Firearm Injuries Drop 20 Percent When Gun Owne...         Biology   
1     Supplementation with probiotics during late pr...          Health   
4     Undisclosed Conflicts of Interests among Biome...  Social Science   
6     One more species of Tardigrade (famed for thei...  Animal Science   
7     There's a20% reduction in gun injuries during ...          Health   
...                                                 ...             ...   
8294  Statements about building walls, deportation a...  Social Science   
8295  Many college students will uncritically accept...  Social Science   
8296  New method for making polymers with perfectly ...       Chemistry   
8297  Research Shows That Doing a Bad Job Wrapping P...  Social Science   
8298  New flame retardants in electronics and childr...     Environment   

                       domain  score  num_comments  sensationalism_score  \
0                    nejm.org     84            22              0.530595   
1           journals.plos.org      8             1              0.482136   
4            ncbi.nlm.nih.gov    263            21              0.511708   
6           journals.plos.org     37             8              0.525921   
7                    nejm.org     14             7              0.478481   
...                       ...    ...           ...                   ...   
8294        journals.plos.org      0             4              0.554148   
8295        journals.plos.org    170            25              0.543118   
8296             pubs.acs.org    148             8              0.503372   
8297  onlinelibrary.wiley.com    175            15              0.589604   
8298             pubs.acs.org      1             0              0.541337   

      jargon_proportion  ... is_top_domain_repo is_top_domain_scam  \
0              0.000000  ...               True              False   
1              0.314286  ...              False              False   
4              0.086957  ...               True              False   
6              0.000000  ...              False              False   
7              0.058824  ...               True              False   
...                 ...  ...                ...                ...   
8294           0.263158  ...              False              False   
8295           0.100000  ...              False              False   
8296           0.171429  ...               True              False   
8297           0.000000  ...              False              False   
8298           0.142857  ...               True              False   

      is_top_domain_unknown  is_top_domain_indecisive  \
0                     False                     False   
1                     False                     False   
4                     False                     False   
6                     False                     False   
7                     False                     False   
...                     ...                       ...   
8294                  False                     False   
8295                  False                     False   
8296                  False                     False   
8297                  False                     False   
8298                  False                     False   

      is_top_domain_less_than_2  label_voting_lm  label_voting_manual  \
0                         False             repo                 repo   
1                         False       scientific           scientific   
4                         False             repo  

In [10]:
inp_df = inp_df[['id', 'title', 'url', 'identifier', 'sem_scholar_title', 'sem_scholar_abstract']]
print(inp_df.shape)
inp_df.head()

(5236, 6)


id                                              title  \
0  811az5  Firearm Injuries Drop 20 Percent When Gun Owne...   
1  814dwi  Supplementation with probiotics during late pr...   
4  815lr6  Undisclosed Conflicts of Interests among Biome...   
6  817tfx  One more species of Tardigrade (famed for thei...   
7  8183ly  There's a20% reduction in gun injuries during ...   

                                                 url  \
0  http://www.nejm.org/doi/full/10.1056/NEJMc1712773   
1  http://journals.plos.org/plosmedicine/article?...   
4       https://www.ncbi.nlm.nih.gov/pubmed/29400625   
6  http://journals.plos.org/plosone/article?id=10...   
7  http://www.nejm.org/doi/full/10.1056/NEJMc1712773   

                         identifier  \
0          DOI:10.1056/NEJMc1712773   
1  DOI:10.1371/journal.pmed.1002507   
4                     PMID:29400625   
6  DOI:10.1371/journal.pone.0192210   
7          DOI:10.1056/NEJMc1712773   

                                   sem_scholar_title  \
0  Reduction in Firearm Injuries during NRA Annua...   
1  Diet during pregnancy and infancy and risk of ...   
4  Undisclosed conflicts of interest among biomed...   
6  An integrative description of Macrobiotus shon...   
7  Reduction in Firearm Injuries during NRA Annua...   

                                sem_scholar_abstract  
0  Decline in Firearm Injuries during NRA Convent...  
1  Background There is uncertainty about the infl...  
4  ABSTRACT Background: Textbooks are a formative...  
6  Tardigrade research in Japan dates back over 1...  
7  Decline in Firearm Injuries during NRA Convent...

In [9]:
inp_df.to_csv('./data/filtered_data_with_abstracts.csv', index=False)

# Conclusion

We were able to get the identifiers for 8299 papers. Out of these, we were able to get _some_ information from semantic scholar API for 6101 papers. Of these, we got 5236 abstracts. 